### PR TITLE
sdk: allow slashtag to accept custom corestores

### PR DESCRIPTION
**Pull request by dzdidi**
Whiledoing something like:
```js
    const slashtag = this.sdk.slashtag(feedId)
    const drive = slashtag.drivestore.get(feedId)
```

I discovered that drives for different `feedIds`  share the same `corestore`. 

This cause issues with deletion of drives as:
```js
   cores = [drive.core, drive.blobs.core] // destroys too little
```
while
```
   cores = Array.from(drive.corestore.cores.values()) // destroys everything
```

### CODE REVIEW

**Comment by Nazeh** [comment]
What do you mean by deletion of drives? @dzdidi 

Using more than one Corestore is not a good idea except in testing. The point of Corestore is to manage storage and Hypercores.

Please explain your goal and use case further.

**Comment by dzdidi** [comment]
Be deletion of drives I mean this operation https://github.com/synonymdev/slashtags-feeds/blob/main/index.js#L180-L211

My current use case is deletion of account data. The problem is that if one creates multiple `slashtag`-`drive` pairs using different `feedId` from the same SDK instance, they all will still share the same core. At the same time deleting cores selected as here https://github.com/synonymdev/slashtags-feeds/blob/main/index.js#L187 is not sufficient. So my choice is deletion of all cores in cores while having separate cores per separate account (also seem to be a good separation).

> Using more than one Corestore is not a good idea except in testing. 

Do you mind elaborate on this?

**Comment by Nazeh** [comment]
@dzdidi You should only create one Slashtag, and keep it the default one, unless you need multiple identities.
```js
const slashtag = sdk.slashtag()
```

Once you have that you can create as many encrypted hyperdrives as you want
```js
const a = slashtag.drive('a')
const b = slashtag.drive('b')
```
These drives absolutely should _not_ be the same, if they are, we have a problem.

Then deleting the underlying storage folders of these drives should be simple: https://github.com/synonymdev/slashtags-feeds/blob/main/index.js#L187

> At the same time deleting cores selected as here https://github.com/synonymdev/slashtags-feeds/blob/main/index.js#L187 is not sufficient.

Why isn't that sufficient?

> Do you mind elaborate on this?

I guess the problem is you were creating multiple slashtags in the first place, Slashtag class is the identity manager, within it you can create multiple drives, no need to create multiple identities.

But also, Corestore is as the name implies is meant to be the one place to manage all hypercores you need, and replicate them all with one command `swarm.on('connection', conn => corestore.replicate(conn))`, If you find yourself thinking I need more than one you are probably wrong and `corestore.session()` or `corestore.namespace()` will serve you better, and Slashtag already does that for you when you call `slashtag.drivestore.get(foo)`, it creates a namespaced session of the corestore so generated cores won't collide with any other namespace, then it uses that session to create hyperdrive:
https://github.com/synonymdev/slashtags/blob/master/packages/drive/index.js#L65

Please try using one Slashtag, and confirm that you can create a unique hyperdrive per unique `feedID`, then let me know why isn't deleting the files of the cores of each drive isn't sufficient to reclaim the disk space.

**Comment by dzdidi** [comment]
> You should only create one Slashtag, and keep it the default one, unless you need multiple identities.

Multiple identities is exactly what I need.

> These drives absolutely should not be the same, if they are, we have a problem.

I am saying that drives' corestore sharing cores (because it is the same corestore)

```js
  async _getDrive(feedId) {
    const slashtag = this.sdk.slashtag(feedId) // same with constant input
    const drive = slashtag.drivestore.get(feedId)
    await drive.ready()

    const cs = [drive.core, drive.blobs.core] // removing from fs destroys too little
    console.log('feed:', feedId)
    cs.map((c) => {
      console.log('drive cores', c.discoveryKey.toString('hex'))
    })
    const cores = Array.from(drive.corestore.cores.values()) // removing from fs  destroys everything
    cores.map((c) => {
      console.log('corestore cores', c.discoveryKey.toString('hex'))
    })
    console.log('----------------------------')
  }


```

```sh
feed: feed_b
drive cores 56a5a477abfd6065ad6f5a616d4cb4610752b12c3c54fdb6bdfd426405030162
drive cores 4e6149988c5176f67cd3e33390ae5c21b3da80dce1ad90a8ddfacd9b325d5090
corestore cores 3ded12d685d13cb1720b0560c7489ffa3e16fd5374a7c08f98dcc31652b85038
corestore cores 56a5a477abfd6065ad6f5a616d4cb4610752b12c3c54fdb6bdfd426405030162
corestore cores 4e6149988c5176f67cd3e33390ae5c21b3da80dce1ad90a8ddfacd9b325d5090
----------------------------
feed: feed_A
drive cores bd6a8d6077dfbe4b9f5796cedf454934be3bc74b7c9a17082c06dc73c15cdbc7
drive cores 00ef1b455f7cea977551bc5384609e94d0362eb9f1753f769fd78a9f51e6c68e
corestore cores 3ded12d685d13cb1720b0560c7489ffa3e16fd5374a7c08f98dcc31652b85038
corestore cores 56a5a477abfd6065ad6f5a616d4cb4610752b12c3c54fdb6bdfd426405030162
corestore cores 4e6149988c5176f67cd3e33390ae5c21b3da80dce1ad90a8ddfacd9b325d5090
corestore cores bacb43155376f30fcb51445b4250094bb2c8694c932359885babbf7c7322b5b7
corestore cores bd6a8d6077dfbe4b9f5796cedf454934be3bc74b7c9a17082c06dc73c15cdbc7
corestore cores 00ef1b455f7cea977551bc5384609e94d0362eb9f1753f769fd78a9f51e6c68e
----------------------------
```

> Why isn't that sufficient?

It selects only last to, but even after removing them I am able to `get` latest value by key.



> I guess the problem is you were creating multiple slashtags in the first place, Slashtag class is the identity manager, within it you can create multiple drives, no need to create multiple identities.

Again, please explain why you think this is a problem.

> But also, Corestore is as the name implies is meant to be the one place to manage all hypercores you need, and replicate them all with one command swarm.on('connection', conn => corestore.replicate(conn)), If you find yourself thinking I need more than one you are probably wrong and corestore.session() or corestore.namespace() will serve you better, and Slashtag already does that for you when you call slashtag.drivestore.get(foo), it creates a namespaced session of the corestore so generated cores won't collide with any other namespace, then it uses that session to create hyperdrive:
https://github.com/synonymdev/slashtags/blob/master/packages/drive/index.js#L65

I do not disagree with you that it is easier to replicate one corestore. However ease of replication is not concern for my use case, while isolation/separation of entities is.

> Please try using one Slashtag, and confirm that you can create a unique hyperdrive per unique feedID, then let me know why isn't deleting the files of the cores of each drive isn't sufficient to reclaim the disk space.

Same issue, because it is the same shared corestore and it is accessible by each drive. See ebove


May you please also confirm that we understand each other on this:

My abstraction layers from top to bottom: 
1) SDK - on instance on server used as a dht client
2) Slashtag - one instance per user on the server (this PR should allow each to have own corestore)
3) Drive - multiple instances per user

Your abstraction layers from top to bottom: 
1) SDK - on instance on server used as a dht client
2) Slashtag - one instance per server
3) Drive - one instances per user

If this is correct, my problems with one suggested by you are:
- with one corestore per server orchestrating drive per user the separation of users only at hyper* stack level (virtual)
- there is a restriction of one drive per user


At the same time one instance of slashtag per user allows one user to have multiple drives per user while each slashtag having own corestore provides isolation on the level of server's file system (physical)

**Comment by Nazeh** [comment]
@dzdidi 
> there is a restriction of one drive per user

One drive per user is exactly what we should aim for, because life gets really hard the more data structures you need to manage and discover over Hyperswarm.

One drive should be all you need to create as many feeds as possible, just put each one in a separate folder, but also, since we are creating an Account feed, and each field is separate, you don't even need more than one feed, but if you do, it is still possible using multiple folders in the drive.

> with one corestore per server orchestrating drive per user the separation of users only at hyper* stack level (virtual)

Good, I promise you the more corestore you have the harder life will be. Also the SDK will need to internally do:
```js
this.swarm.on('connection', conn => {
   this.corestores.forEach((store) => { store.replicate(conn })
})
```
Or you will have to do that manually for every corestore:
```js
sdk.swarm.on('connection', conn => corestore.replicate(conn))
```

**Comment by Nazeh** [comment]
If you absolutely need to create separate drive per context per user do this:

```js
const foo = slashtag.drivestore.get('userA/foo')
const bar = slashtag.drivestore.get('userA/bar')
```

But Slashtag is basically a Signer, a writer, no need to have more than one per server.

File: packages/sdk/index.js [hunk 1]
```diff
@@ -90,18 +90,23 @@ export class SDK extends EventEmitter {
   /**
    * Creates a Slashtag by name.
    * @param {string} [name] utf8 encoded string
+   * @param {object} [opts] slashtag instantiation options
+   * @param {Corestore} [opts.corestore] corestore
+   * @param {DHT} [opts.dht] dht
```
Comment (Nazeh) [inline packages/sdk/index.js:95]: Why do you need to pass a custom DHT too? If you are going to use a custom corestore + custom DHT what do you need the SDK for? Just to generate drive keys?

If so then, it might be a better idea to separate these key generation functionality into their own methods, so you can access them and do with them what you like.

For example:

```js
const { auth, encryptionKey } slashtag.drivestore.generateOptions(feedID)

const drive = new Hyperdrive({ auth, encryptionKey })
```

Comment (dzdidi) [reply packages/sdk/index.js:95]: I am simply allowing use of SDK to have same amount of customization as with the direct use of Slashtag.

**Comment by dzdidi** [comment]
@Nazeh  is there any other problem than replication with my design?

**Comment by Nazeh** [comment]
@dzdidi Yes, giving users options to shoot themselves in the foot is not something we should do in this SDK.

If you really need to use external Corestores and DHTs, then you can easily do that like this:

```js
const drive = slashtag.drivestore.get(feedID)
await drive.ready()
const myDrive = myCorestore.get({ auth: drive.core.auth, encryptionKey: drive.core.encryptionKey })
await myDrive.ready()
drive.close()
```

Or we can expose a new method that returns the `auth` and `encryptionKey` without actually creating the `Hyperdrive`, which should be easy to do.

**Comment by dzdidi** [comment]
@Nazeh Excluding the context of intended usage (which is fine for open source SDK intended for general public), the proposed change itself is about "allowing use of SDK to have same amount of customization as with the direct use of Slashtag.". What is the problem with doing this?

**Comment by Nazeh** [comment]
@dzdidi The problem is that it defeats the purpose of the SDK (best practices, automating tedious boilerplates, and sensible defaults) ... if you want customization you should use Holepunch's modules directly.

It also makes maintaing this SDK much harder, there is already too moving parts and degrees of freedoms internally, and testing all these edge cases are hard. Adding more options doesn't help.

The only exception is getting Slashtag's key generation, which you can already get as described above, but also can be provided as a separate dead simple module if there is demand for that.

**Comment by dzdidi** [comment]
> The problem is that it defeats the purpose of the SDK (best practices, automating tedious boilerplates, and sensible defaults) ... 

This change does not change any defaults, thus I do not see how does it lead to distortion of best practices or boilerplates

> if you want customization you should use Holepunch's modules directly.

disagree, one should use Holepunch's modules directly only if they want to use Holepunch's modules. Ideally Slashtag should be agnostic from the underlying data transfer protocol which in perfect world should be plugable via interface defined by Slashtag

> It also makes maintaing this SDK much harder, there is already too moving parts and degrees of freedoms internally, and testing all these edge cases are hard. Adding more options doesn't help.

I agree that each change involves adding test coverage for it, but this particular change allows functionality already available in Slashtags thus it is mostly covered by its tests

**Comment by Nazeh** [comment]
@dzdidi Passing Corestore and DHT are not "agnostic from the underlying data transfer protocol" though, the are Holepunch modules, so more reason to keep them abstracted from the user.

> I agree that each change involves adding test coverage for it, but this particular change allows functionality already available in Slashtags thus it is mostly covered by its tests

No, it overrides Corestore and DHT, without guaranteeing that peers are going to replicate correctly with that Corestore, or that this Corestore is the right version at all, or have all the APIs that are used internally.

I don't see the point of encouraging this risky mode of using the SDK.

**Comment by dzdidi** [comment]
@Nazeh , 
> Passing Corestore and DHT are not "agnostic from the underlying data transfer protocol" though, the are Holepunch modules

I've said "**Ideally** Slashtag **should be** agnostic ... **in perfect world** ..." . Which also does not contradict to Holepunch being used now "as hardcoded defaults".


> No, it overrides Corestore and DHT, without guaranteeing that peers are going to replicate correctly with that Corestore, or that this Corestore is the right version at all, or have all the APIs that are used internally.

it **allows** to overwrite defaults, there are no change to defaults in this PR.

> I don't see the point of encouraging this risky mode of using the SDK.

There is not encouragement but **allowance**.  

While defaults should be most safest and convenient options, ability to change them should not be stripped away of from the user. 

While change of default settings always comes with risk only user can bare. If you think there should be some kind of a warning regarding consequences of changing defaults I am happy to add it to readme or other place

**Comment by Nazeh** [comment]
For better or for worse everything is allowed in JS:
```js
slashtag.drivestore.corestore = myCorestore;
slashtag.dht = myDHT
```

But public APIs are hints, endorsements, and a responsibility to maintain, which we should not do in this case I submit.

**Comment by dzdidi** [comment]
May you please elaborate on reasoning, because so far everything I was able to extract from this conversation is that custom corestore will require its explicit replication

**Comment by Nazeh** [comment]
As I said before Corestore is meant to be a singleton that manages all cores in a single application, using multiple ones is just a sign that the user is misunderstanding its purpose, and there is always a better way to achieve what they are trying to achieve.

Moreover, the SDK expects any drive created using itself to be saved in the same storage directory, and the data to be accessible to the internal Corestore. If that wasn't the expectation, there would have been no reason to use the Corestore in the first place.

**Comment by dzdidi** [comment]
> As I said before Corestore is meant to be a singleton that manages all cores in a single application, using multiple ones is just a sign that the user is misunderstanding its purpose, and there is always a better way to achieve what they are trying to achieve.

I am asking for reasoning behind why "Corestore is meant to be a singleton"

> Moreover, the SDK expects any drive created using itself to be saved in the same storage directory, and the data to be accessible to the internal Corestore. If that wasn't the expectation, there would have been no reason to use the Corestore in the first place.

I do not understand the purpose of drive inside of SDK but outside of slashtag?

**Comment by Nazeh** [comment]
The SDK should never support more than a single Corestore because that defeats the purpose of having a Corestore in the first place, we could have just juggled Hypercores ourselves inside the SDK. Again 2 Corestores is like using two databases in your app instead of creating two tables.

You do need drives inside the SDK but outside of Slasthag, to read remote Drives, that might not belong to a Slashtag in the first place, as in Feeds, but also because there is very little utility to create a read only Slashtag for your contact, only to read their public drive, instead you just use `SDK.drive(contactPublicKey)`

**Comment by BitcoinErrorLog** [comment]
@dzdidi Can you please overview the issue and the result you want? 

Generally, afaik, Ar is correct that there should probably only be one Corestore that routes to a collections of drives. 

Are you saying you are missing a way to delete a single drive under the Corestore, or are you saying you have a use case for many Corestores per Slashtag, or what?

The intended design for Slashtags is one user may have drives about their own key and drives about other keys. In the case of a website they would mostly be concerned with drives about their users. The website would only be using one Slashtag keypair to do this, with a subordinate drive about each user beneath that. 

I'm not sure if I am helping, but I want to make sure this work results in a consistent design, without ignoring whether there is a use case missing here.

**Comment by dzdidi** [comment]
> You do need drives inside the SDK but outside of Slasthag, to read remote Drives, that might not belong to a Slashtag in the first place, as in Feeds, but also because there is very little utility to create a read only Slashtag for your contact, only to read their public drive, instead you just use SDK.drive(contactPublicKey)

@Nazeh  got it, thanks

@BitcoinErrorLog 

> Are you saying you are missing a way to delete a single drive under the Corestore,

yes

> or are you saying you have a use case for many Corestores per Slashtag, or what?

I do not have any in mind but this change may enable them

@Nazeh 

> Deleting a single drive or a single core from the storage doesn't really have easy API (yet), but it is not hard, and we do it already here https://github.com/synonymdev/slashtags-feeds/blob/main/index.js#L186

As stated in the description this deletion does not work, I am still able to read data.


The work around I've found is to have one corestore per user as this will allow delete entire corestore for its user also without risk to affect other users.


> It actually gets much harder to delete drives and cores, when you don't know where they were stored because you created them with a custom Corestore and now you don't remember where it was stored on disk.

there are ways to "memorize" input parameters to functions.


It seems like I completely misaligned on the value prop of SDK. I think it is something we should build our entire stack around. While we have opposite points of view on SDK, I do not think it is a reason this discussion reached this length. It is my failure to understand Ar's  position from technical standpoint of why having multiple corestores per sdk is bad idea. I acknowledge argument that it breaks automatic replication which is fine in my opinion. But this seem to be the only technical argument I've received. As for "it does not make sense" in my opinion the fact that this change enables deletion is sufficient to say that it does make sense. 







Still, all my questions about design decisions taken at SDK are open, the only technical reasoning I've read so far is "one corestore make replication easier".

**Comment by Nazeh** [comment]
> As stated in the description this deletion does not work, I am still able to read data.

Depending on your code (I didn't see it), you probably are reading from [cached blocks inside Hypercore (memory)](https://github.com/holepunchto/hypercore/blob/master/index.js#L56), but if you remove the folders, you remove the folders, disk space is freed. If not, then I need to see the code.

**Comment by dzdidi** [comment]
> Depending on your code (I didn't see it), you probably are reading from [cached blocks inside Hypercore (memory)](https://github.com/holepunchto/hypercore/blob/master/index.js#L56), but if you remove the folders, you remove the folders, disk space is freed. If not, then I need to see the code.

Thanks for this one, is there a way to disable or skipping caching form SDK's API? Quick search did not give me anything

**Comment by Nazeh** [comment]
@dzdidi No https://github.com/holepunchto/hypercore/blob/master/index.js#L711 Hypercore doesn't expose that API

So your best shot to test is either call [`drive.core._get`](https://github.com/holepunchto/hypercore/blob/789527d9155b69c3d921b8f06d82e4bb42173c8a/index.js#L741), or just check the disk like [slashtags feed does](https://github.com/synonymdev/slashtags-feeds/blob/6bc87090c1d318b6ea13bd5bdf335ba89c2de030/test/all.js#L207)

**Comment by dzdidi** [comment]
@Nazeh thank you

> No https://github.com/holepunchto/hypercore/blob/master/index.js#L711 Hypercore doesn't expose that API

At least there is an option to instantiate hypercore without cache, correct? But I do not see the way to do this with SDK, right?

> So your best shot to test is either call [drive.core._get](https://github.com/holepunchto/hypercore/blob/789527d9155b69c3d921b8f06d82e4bb42173c8a/index.js#L741), or just check the disk like [slashtags feed does](https://github.com/synonymdev/slashtags-feeds/blob/6bc87090c1d318b6ea13bd5bdf335ba89c2de030/test/all.js#L207)

The last option reads folders on fs, they are removed I confirm. My point is that some core folders are shared between drives within same corestore (see console output https://github.com/synonymdev/slashtags/pull/83#issuecomment-1441300771) in append only manner. Also notice that the very first core dir does not get selected into `drive.core, drive.blobs.core` and I could not find where it can be picked from except for corestore. My assumption is that it is exactly the core we do not delete while should but again I could not find anywhere except corestore.

As for the first option, I'll try but not sure what it can give me. If I understand correctly it reads from core. Calling it on removed cores surely will return nothing (or throw an error) so there is no meaning in that. Calling on the core may prove that value is there either in cache or not so as a result it will tell either that value exists in cache which I have no option to invalidate or skip or even create storage without cache unless doing it manually. Or that it exists in file and my initial assumption that we do not remove all necessary cores is correct. In either way, the end solution to this removal proble is removing entire corestore. 

Does it make sense or there is some fallacy in here?

**Comment by Nazeh** [comment]
The SDK can't expose an option to create drives without cache see https://github.com/holepunchto/hyperdrive-next/blob/master/index.js#L461 but even if we can, not much upside for that.

All corestores in all drives are the same corestore, with just different namespace so they generate different cores for the same name, see https://github.com/holepunchto/corestore/blob/master/index.js#L315

So it makes sense you should see all discoveryKeys in the same Corestore.

As for the first core that you don't know what it is, it is an internal database in the SDK to keep track of all the names we used so far, just to make it easier to list them https://github.com/synonymdev/slashtags/blob/master/packages/drive/index.js#L19

See https://github.com/synonymdev/slashtags/blob/master/packages/drive/index.js#L30 which allows to list drives created so far, not just random cores, since Corestore also doesn't keep a list of the names used to create each core.

**Comment by dzdidi** [comment]
> The SDK can't expose an option to create drives without cache see https://github.com/holepunchto/hyperdrive-next/blob/master/index.js#L461 but even if we can, not much upside for that.

I see, thanks

> All corestores in all drives are the same corestore, with just different namespace so they generate different cores for the same name, see https://github.com/holepunchto/corestore/blob/master/index.js#L315
So it makes sense you should see all discoveryKeys in the same Corestore.

got it

> As for the first core that you don't know what it is, it is an internal database in the SDK to keep track of all the names we used so far, just to make it easier to list them https://github.com/synonymdev/slashtags/blob/master/packages/drive/index.js#L19

thanks

> See https://github.com/synonymdev/slashtags/blob/master/packages/drive/index.js#L30 which allows to list drives created so far, not just random cores, since Corestore also doesn't keep a list of the names used to create each core.

Not sure I understand this one, I need to play with it


Overall do I understand correctly you are saying that it must be cache issue I am hitting that returns deleted value?

If this is the case, it leaves me wondering why removing specific core does not remove cache while removing entire corestore does. 

Would it make sense to try remove cores of a drive and then reload corestore to verify if drives were removed?

**Comment by Nazeh** [comment]
Again can't judge while not seeing the code, but deleting data from disk will do the job, I assure you.

My theory though is that deleting the whole corestore directory causes an issue with the corestore lock, and some error is thrown somewhere and some error catcher reacts to free memory.

I don't even care about deleting cores, not until there is data at all to worry about desk space.

Once we are there, better more lawful ways to clear the disk are easy to add.

I know all of this s confusing, sorry for not being more clear and helpful.

**Comment by Nazeh** [comment]
I think you should use https://github.com/synonymdev/slashtags-feeds/
We already have multiple Feeds running using that module. If we want that module to use the SDK and not just Hyperdrive and Hyperswarm directly, we can do that there.

Either use slashtags-feeds as is, or open a PR changing it to use the SDK, and we can discuss that change there.

**Comment by Nazeh** [comment]
This has been eye-opening though, the SDK's api is not near as good as I wanted it to be, mostly because it tries to be a thin wrapper around Holepunch's modules.

Going forward, I hope we can build things differently and be more assertive to reach the API users expect with as little unexpected weirdness as possible.

That will require a lot of breaking changes, but hey, we are still in `alpha` :)

Closing this for now.

**Comment by dzdidi** [comment]
> Either use slashtags-feeds as is, or open a PR changing it to use the SDK, and we can discuss that change there.

I see that slashtags-feeds was using SDK before, what was the reason to move away from it?

**Comment by Nazeh** [comment]
The SDK was changing too much too often, that module was intended to be used by outsiders much more than the close integration between the SDK and Bitkit, so I wanted to keep it shielded from the SDK.

Another reason, is that it will be a server side module anyways, while the SDK is more useful for cross platform and identity management in a wallet or a web app.

A third is that feeds are such a separate feature from identity right now, that the SDK can entirely change its stack, while feeds keep using Hyperdrive.

And the fact that it works without the SDK shows that it doesn't need it much for now.

Basically I was and still would like to have less stuff dependant on the SDK as possible to have some freedom to iterate in the service of Bitkit
